toolkit/tools/pkg/imagecustomizerlib/customizeoverlays.go

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package imagecustomizerlib
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/microsoft/azurelinux/toolkit/tools/imagecustomizerapi"
+	"github.com/microsoft/azurelinux/toolkit/tools/imagegen/diskutils"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/logger"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/shell"
+)
+
+func enableOverlays(
+	overlays *[]imagecustomizerapi.Overlay, selinuxMode imagecustomizerapi.SELinuxMode,
+	fileSystems []imagecustomizerapi.FileSystem, imageChroot *safechroot.Chroot) (bool, error) {
+	var err error
+
+	if overlays == nil {
+		return false, nil
+	}
+
+	logger.Log.Infof("Enable filesystem overlays")
+
+	// Integrate overlay driver into the initramfs img.
+	overlayDriver := "overlay"
+	err = addDracutDriver(overlayDriver, imageChroot)
+	if err != nil {
+		return false, err
+	}
+
+	// Dereference the pointer to get the slice
+	overlaysDereference := *overlays
+	err = updateFstabForOverlays(overlaysDereference, fileSystems, imageChroot)
+	if err != nil {
+		return false, fmt.Errorf("failed to update fstab file for overlays:\n%w", err)
+	}
+
+	// Create necessary directories for overlays
+	err = createOverlayDirectories(overlaysDereference, imageChroot)
+	if err != nil {
+		return false, fmt.Errorf("failed to create overlay directories:\n%w", err)
+	}
+
+	// Add equivalency rules for each overlay
+	err = addEquivalencyRules(selinuxMode, overlaysDereference, imageChroot)
+	if err != nil {
+		return false, fmt.Errorf("failed to add equivalency rules for overlays:\n%w", err)
+	}
+
+	return true, nil
+}
+
+func updateFstabForOverlays(
+	overlays []imagecustomizerapi.Overlay, fileSystems []imagecustomizerapi.FileSystem,
+	imageChroot *safechroot.Chroot) error {
+	var err error
+
+	fstabFile := filepath.Join(imageChroot.RootDir(), "etc/fstab")
+	fstabEntries, err := diskutils.ReadFstabFile(fstabFile)
+	if err != nil {
+		return fmt.Errorf("failed to read fstab file: %v", err)
+	}
+
+	var updatedEntries []diskutils.FstabEntry
+	updatedEntries = append(updatedEntries, fstabEntries...)
+
+	for _, overlay := range overlays {
+		lowerDirs := overlay.LowerDirs
+		upperDir := overlay.UpperDir
+		workDir := overlay.WorkDir
+		mountDependencies := overlay.MountDependencies
+
+		// Validate that each mountDependency has the x-initrd.mount option in
+		// the corresponding fstab entry.
+		for _, dep := range mountDependencies {
+			found := false
+			for _, entry := range fstabEntries {
+				if entry.Target == dep {
+					found = true
+					if !strings.Contains(entry.Options, "x-initrd.mount") {
```

Comment:
It would be goodness to split the options by `,` and then check if the `x-initrd.mount` option is present. This is more robust, in case there is ever a new option that includes `x-initrd.mount` in its name.